### DISCUSSION TITLE
Preferred way of saving "settings" in an Avalonia project

### DISCUSSION THREAD

**Discussion by duck-dev** [AvaloniaUI/Avalonia#7190] [Q&A]
Hello everyone,

I was wondering what your preferred way of saving custom settings in an Avalonia project is. Let's say, I want to save a few properties, like a preferred theme, a path, a specific colour etc. In WPF, you'd have the possibility to use [Application Settings](https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2015/ide/managing-application-settings-dotnet?view=vs-2015&redirectedfrom=MSDN) ([User Settings](https://docs.microsoft.com/en-us/archive/blogs/patrickdanino/user-settings-in-wpf)), but that doesn't seem to be possible with an Avalonia project, at least not natively. 
What are your experiences with Avalonia and saving settings? Is it maybe better to just save my own XML/JSON file somewhere, read and write to it?

I wasn't able to find anything and for WPF, people seem to suggest the mentioned Application Settings only, which are natively supported in a WPF project.

**Comment by maxkatz6** [comment]
Unrelevant to Avalonia. 
Save it as you would save it with any other .net core app. Your options at least:
1. Json/xml file
2. Maybe Microsoft.Extensions.Configuration will work (basically the same as json, but with abstraction)
3. Local database (litedb? sqlite?)
4. Web service
5. OS specific storage, if available (NSUserDefaults on macos, MSIX ApplicationData on windows, I haven't used those). 
6. ???

Remember that Avalonia is a GUI framework, and it doesn't limit you in packaging or features like this. 
It's up to you to decide, what's the best in your situation.

**Reply by duck-dev** [reply]
I am totally aware of that. My question was rather directed to a good way fitting the architecture (MVVM). There are certain things that "work" better with certain architectures and frameworks.

But I already assumed a custom way would be a good way to go.

Thank you for your answer

**Reply by cyraid** [reply]
I agree, it'd be nice for an Avalonia way, as different OS's have different paths, etc. They already have IStorageProvider. Plus people are constantly having to rewrite their settings solution every time they create a new project. Definitely would be nice to have one there as it's so common.

**Reply by LaurentInSeattle** [reply]
Already done, for all platforms supporting .Net, see here:   https://learn.microsoft.com/en-us/dotnet/core/extensions/configuration

**Reply by cyraid** [reply]
@LaurentInSeattle if I'm not mistaken, IConfiguration is mostly a one way street. You'd still need to implement dumping it all to the file.

**Reply by mysteryx93** [reply]
.NET configuration is designed for ASP.NET configuration mostly: pulling settings from various sources to configure your service. A desktop app, on the other hand, needs to store user settings somewhere in a more interactive way into persistent storage. For an ASP.NET website, there may also be user settings to store, but they'll be stored in a database for the user. .NET Configuration is used to configure access to that database and more architectural settings like that.

**Comment by mysteryx93** [comment]
Take a look at my [AppSettingsProvider](https://github.com/mysteryx93/HanumanInstituteApps/blob/master/Src/App.Player432Hz/Player432Hz/Business/AppSettingsProvider.cs) and [SettingsProviderBase](https://github.com/mysteryx93/HanumanInstituteApps/blob/master/Src/Services/Utilities/SettingsProviderBase.cs)

**Comment by Hunv** [comment]
As this is the first result when you search for saving settings in Avalonia and the two provided links are not available anymore:
Are there any recommendations? MAUI has the Preferences class to do that on each OSes default user storage.
Is there something similar that can be used when using Avalonia nowadays (doesn't matter if 3rd party or not)?
I understand that this is not UI so it is not directly related to Avalonia but if you are coming from Xamarin or MAUI you will have this question sooner or later.

**Reply by thevortexcloud** [reply]
There is nothing stopping you from using Xamarin/MAUI essentials with Avalonia. You can also just straight up use MAUI and Avalonia side by side, it's an officially supported scenario.

https://github.com/AvaloniaUI/AvaloniaMauiHybrid

Using Xamarin Essentials is also how most people are handling things like mobile APIs with Avalonia. Do keep in mind MAUI/Avalonia/Xamarin do not support all of the same platforms (MAUI notably does not have any official support for Linux), so some things may not work in some places.

That said, it's generally fairly simple to create a simple abstraction layer for key value storage. For desktops you can just use a file/SQLite database, most other platforms have some sort of built in database system you can use.

**Reply by mysteryx93** [reply]
I updated the links to my solution. Note that my solution hasn't taken mobile devices into account yet.

**Reply by Hunv** [reply]
Thank you. Too bad, I need Mobile Devices and Desktop.

**Reply by mysteryx93** [reply]
On mobile, can you store a local XML file on the device in a writeable folder just like on desktop? Or do you need to use special user storage? Either way, the only difference would be "where" you write the XML. Should work just fine I think.

That's good for apps with simple settings, where you rewrite the entire file to edit something. It has the benefit of caching the settings and exposing them to the apps without reading on the disk each time.

If you have hundreds of settings and want to alter any of them on-demand, you might need a different approach (key-value storage).

**Reply by Hunv** [reply]
Yes. The nice and easy thing on things like the Preferences class is, that you don't have to take care where it needs to be stored etc. It is always in the plattforms users "drive" and on the right place.
Doing it on my own means that, next to the save/load functions itself, I have to figure out what the correct paths are, what are limitations etc. and handle them on my own. That was the reason why I asked before I need to start inventing the wheel again now.
But I understand, that Avalonia is UI only and does not provide this kind of "helpers". That is sad, but I fully understand it.

**Reply by mysteryx93** [reply]
Having too much black magic isn't good either. You want to be in control of your data and know where it's stored.

Like if you want to import/export your settings file. IMO there's nothing worse than apps where you can't figure out where they store their stuff and you want to backup your settings and data.

**Reply by ptasznikarium** [reply]
Following is a helper class that I wrote for one of my projects. You can inject it via DI or change to be static class if you prefer.
Although I wanted to make it universal, so far I only tested on Windows, so keep it in mind.

```csharp
public class FileSystemPaths : IFileSystemPaths
{
    private const string _cache_Folder = "cache";
    private const string _logs_Folder = "logs";

    private readonly Lazy<string> _userDataPath;
    private readonly Lazy<string> _cachePath;
    private readonly Lazy<string> _logsPath;

    private readonly IAppInfo _appInfo;

    public FileSystemPaths(IAppInfo appInfo)
    {
        _appInfo = appInfo;

        _userDataPath = new Lazy<string>(() => GetUserDataPath());
        _cachePath = new Lazy<string>(() => Path.Combine(_userDataPath.Value, _cache_Folder));
        _logsPath = new Lazy<string>(() => Path.Combine(_userDataPath.Value, _logs_Folder));
    }

    public string UserDataPath => _userDataPath.Value;
    public string CachePath => _cachePath.Value;
    public string LogsPath => _logsPath.Value;

    private string GetUserDataPath()
    {
        // Windows:
        // ApplicationData: C:\Users\<user>\AppData\Roaming
        // LocalApplicationData: C:\Users\<user>\AppData\Local
        if (OperatingSystem.IsWindows())
        {
            return Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                _appInfo.CompanyName, _appInfo.AppName);
        }
        // Linux:
        // ApplicationData: /home/<user>/.config - respects custom XDG_CONFIG_HOME
        // LocalApplicationData: /home/<user>/.local/share - respects custom XDG_DATA_HOME
        // macOS:
        // both resolve to: /Users/<user>/Library/Application Support
        else if (OperatingSystem.IsLinux() || OperatingSystem.IsMacOS())
        {
            return Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
                _appInfo.CompanyName, _appInfo.AppName);
        }
        // Mobile platforms use sandboxed storage, no need for CompanyName/AppName subfolders
        // Android sandbox:
        // both resolve to: /data/user/0/<package.name>/files
        // iOS sandbox:
        // both resolve to: /var/mobile/Containers/Data/Application/<GUID>/Library/Application Support
        else if (OperatingSystem.IsAndroid() || OperatingSystem.IsIOS())
        {
            return Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
        }
        // Unsupported platform
        else
        {
            throw new PlatformNotSupportedException($"Unsupported Operating System ({Environment.OSVersion.VersionString}).");
        }
    }
}
```

**Comment by sb-viktor** [comment]
Is there any implementation for a project that shows how to create, resave and use settings? I don't see anything wrong with Avalonia not offering a specific solution, it's probably a good thing (use whatever you want). But it creates some problems. Creating modifiable settings is very complicated and time-consuming (especially if the application is large).

### JSON
I don't know how it looks like, if I understand correctly, the application parses JSON every time it is launched. I basically don't like the fact that all program settings are stored in a separate file that can be changed by anyone. Here are some problems that arise in this case:

1. File loss ( You will have to handle a lot of exceptions )
2. broken JSON ( unlikely, but cannot be excluded )
3. insecurity ( Users can manually edit the file, which may cause the application to work incorrectly )
4. Lack of typing ( JSON does not support complex data types "out of the box", you have to serialize/deserialize objects manually )
5. Problems with versioning (When changing the structure of settings in a new version of the application, conflicts with old files may occur)
6. No validation
7. Difficulty separating settings ( For example, if you need to differentiate settings for different users or environments )

All( almost everything ) the same can be said about XML

### DIY
It's actually not a bad option, if you don't consider it labor-intensive. In the end, you will have to handle all errors, all parameter updates, all exceptions yourself

### What Else
Nothing really. I don't think that databases are suitable for storing settings ( they are certainly can, but about the same as JSON. That is, possible, but there are a lot of problems). And OS specific storage deprives the project of its main advantage - cross-platform.

**Reply by thevortexcloud** [reply]
> all program settings are stored in a separate file that can be changed by anyone. 

This is always possible as long as settings are saved locally. It's just a question of how obtuse you make it. As a user, my settings are **_my_** settings and I want to be able to change them however I want.

> File loss

You can just regenerate the file. It's not hard.

> broken JSON 

Display an error message and close. It's user error at that point.

> insecurity

If you have security sensitive stuff in a persisted JSON file you are probably doing something wrong.

> Lack of typing ( JSON does not support complex data types "out of the box", you have to serialize/deserialize objects manually )

Not an issue with C#. The JSON serialiser does it all for you. Some serialsiers even support polymorphic serialisation.

> Problems with versioning (When changing the structure of settings in a new version of the application, conflicts with old files may occur)

Version your file. If the version does not match, upgrade it. In any case, this is really not an issue unless you start changing data types. Old values are typically ignored and new ones can be made optional.

> No validation

You can add your own very easily. If you mean validating the JSON file itself, the the JSON library you use will do that for you.

> Difficulty separating settings ( For example, if you need to differentiate settings for different users or environments )

You can just use multiple files? The .NET config library for example allows you to register a secret JSON file which is stored outside your project directory, but at runtime gets merged into the main settings file.

https://learn.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-9.0

I am actually not clear why this is a problem. 

> It's actually not a bad option, if you don't consider it labor-intensive. In the end, you will have to handle all errors, all parameter updates, all exceptions yourself

It takes ~2 hours (sometimes less) to make a basic configuration system using what ever data store you want. I have done it myself many times. I will sometimes even make multiple for the same project. EG I will make an `IConfiguration` interface which is implemented by an SQLite database reader and also a file reader. As fiddling with a text file is a lot easier and faster than a database during development. Then I just swap implementations as needed.

Anyway, the same complaints you list apply to basically _all_ code you will ever write.

> Nothing really. I don't think that databases are suitable for storing settings 

Why? Databases are great for key value storage. They are way faster than parsing and updating a JSON file and often more portable too. SQLite is also extremely lightweight and does not even require running a server.

If you want to make things even more annoying for people though, just make a binary file (make sure your endianess is correct though).

**Reply by sb-viktor** [reply]
Well @thevortexcloud, I understand your point. Implementation in _~2 hours (sometimes less)_ is possible only in one case: you've done it before and just migrate the feature from one project to another. That's why I asked if there is some kind of template like those in [AvaloniaUI.QuickGuides](https://github.com/AvaloniaUI/AvaloniaUI.QuickGuides). Program settings are probably what every application needs, so I thought some sort of template should exist.

Sort of like in WPF there was a much more convenient option, where you just set the name, data type and could then change it, save it, etc. And everyone had a Default property, and with a command like `Application.Settings.Default()` it was reset to factory defaults.

***I say "sort of" because it's been a long time since I worked with WPF and I don't remember if this configuration was a separate file or not**

Anyway it sounds easier than creating a database ( if I understand correctly how it works )

**Reply by thevortexcloud** [reply]
> I understand your point. Implementation in ~2 hours (sometimes less) is possible only in one case: you've done it before and just migrate the feature from one project to another. 

Nope. I always do it from scratch and that's about how long it takes me if I am trying to do it semi properly. If I just want it done fast, I can hack it together in ~15 minutes or less.  It would honestly be slower for me to set up packages to reuse it. Seriously can't get much simpler than ~5 lines of C# to generate a JSON file from C# code. EG this is a _very_ simple implementation (which may or may not work, I have not tried it)

```csharp
public sealed record  Settings {
public string MySetting { get; init ;}
}

public sealed class SettingsManager {
    public async Task SaveAsync(Settings settings) {
    using (var fs = new FileStream(settingsPath, FileMode.Open)) {
      await  JsonSerialiser.SerialiseAsync(settings, fs);
    }
  }
  
  public async Task Get<Settings>() {
      using (var fs = new FileStream(settingsPath, FileMode.Open)) {
        await  JsonSerialiser.DeserialiseAsync<Settings>(fs);
      }
  }
}
```
> Anyway it sounds easier than creating a database ( if I understand correctly how it works )

Depends on what you are doing. If you need to handwrite SQL queries, then it can be slower. But there are lots of llibraries like dapper/EF Core available which make it much easier. Databases do have the advantage of generally being ACID compliant though. Random files, not so much.

**Comment by xmaxrayx** [comment]
text is the best but u need write ur class and manage read\writings 

json is to bloated and need more cpu to parse
DB are 20-50mb DLL to load just to open 1kb file....

**Reply by mysteryx93** [reply]
> json is to bloated

I can't think of a use-case where json is too bloated. If json is too bloated, .NET itself is too bloated.

but If you want fastest performance in exchange for readable settings file, use a binary serializer like [MemoryPack](https://github.com/Cysharp/MemoryPack)

**Reply by xmaxrayx** [reply]
@mysteryx93 

`.Net` can get better like recent stuff with unmanaged support `,json` in another hand is "standard"  u dont have any power to improve it other than the parsing.

u can write struct or memory to file without 3rd party lib.

**Reply by thevortexcloud** [reply]
Actually, JSON support is now part of `System.Text`. It even has source generator support for AOT compilation and avoiding reflection. It is however still shipped as a separate package. But if you are using pretty much any other modern package, there is a fairly decent chance you are going to need to include it anyway as a dependency.

> u can write struct or memory to file without 3rd party lib.

You can, but this has a lot of downsides:

1. The user can't easily edit a binary file
2. The file requires special tools to inspect
3. The file likely wont work properly if you have changes to your internal structures
4. You need to deal with endianness (which is something most people are not even aware is a thing)
5. There are potential security problems.  EG the BinaryFormatter has been completely disabled because of unfixable security problems. https://learn.microsoft.com/en-us/dotnet/standard/serialization/binaryformatter-security-guide

If you really want to avoid any packages, I am pretty sure XML works out of the box in .NET?

**Reply by mysteryx93** [reply]
Really -- no need to re-invent the wheel

**Comment by dabblob** [comment]
I use Binary file with protobuf-net it's fast and easy to use
but you will need to handle File IO
